internal/handlers/common/params.go

```diff
@@ -99,104 +100,76 @@ func GetWholeNumberParam(value any) (int64, error) {
+			}
+
+			if b < 0 {
+				return 0, NewError(ErrBadValue, fmt.Errorf("bit positions must be >= 0 but got: %d: %d", i, b))
+			}
+
+			bitmask |= 1 << uint64(math.Min(float64(b), 63))
 		}
-
 	case int32:
 		// {field: {$bitsAllClear: bitmask}}
-		if value < 0 {
-			return types.Binary{}, errNegativeNumber
+		if mask < 0 {
+			return 0, errNegativeNumber
 		}
 
-		mask, err = types.BinaryFromInt(int64(value))
-		if err != nil {
-			return types.Binary{}, NewError(ErrBadValue, err)
-		}
+		bitmask = uint64(mask)
 	case int64:
 		// {field: {$bitsAllClear: bitmask}}
-		if value < 0 {
-			return types.Binary{}, errNegativeNumber
+		if mask < 0 {
+			return 0, errNegativeNumber
 		}
 
-		mask, err = types.BinaryFromInt(value)
-		if err != nil {
-			return types.Binary{}, NewError(ErrBadValue, err)
-		}
+		bitmask = uint64(mask)
+
 	case float64:
+		// {field: {$bitsAllClear: bitmask}}
 		// TODO check float negative zero
-		if value != math.Trunc(value) || math.IsNaN(value) || math.IsInf(value, 0) {
-			return types.Binary{}, errNotWholeNumber
+		if mask != math.Trunc(mask) || math.IsNaN(mask) || math.IsInf(mask, 0) {
+			return 0, errNotWholeNumber
 		}
-		mask, err = types.BinaryFromInt(int64(value))
-		if err != nil {
-			return types.Binary{}, err
+
+		if mask < 0 {
+			return 0, errNegativeNumber
 		}
-	case types.Binary:
-		// {field: {$bitsAllClear: BinData()}}
-		mask = value
-	default:
-		return types.Binary{}, errNotBinaryMask
-	}
-	return mask, nil
-}
 
-// getBinaryParam matches value type, returning types.Binary and error if match failed.
-func getBinaryParam(value any) (types.Binary, error) {
-	var res types.Binary
-	var err error
+		bitmask = uint64(mask)
 
-	switch value := value.(type) {
-	case int32:
-		res, err = types.BinaryFromInt(int64(value))
-		if err != nil {
-			return types.Binary{}, err
-		}
-	case int64:
-		res, err = types.BinaryFromInt(value)
-		if err != nil {
-			return types.Binary{}, err
-		}
-	case float64:
-		// TODO check float negative zero
-		if value != math.Trunc(value) || math.IsNaN(value) || math.IsInf(value, 0) {
-			return types.Binary{}, errNotWholeNumber
-		}
-		res, err = types.BinaryFromInt(int64(value))
-		if err != nil {
-			return types.Binary{}, err
-		}
 	case types.Binary:
-		return value, nil
-	default:
-		return types.Binary{}, NewErrorMsg(ErrBadValue, "not matched")
-	}
-	return res, nil
-}
-
-// getBinaryParams creates types.Binary for field and mask from given values returning types.Binary or error.
-func getBinaryParams(fieldValue any, maskValue any) (types.Binary, types.Binary, error) {
-	maskBinary, err := getBinaryMaskParam(maskValue)
-	if err != nil {
-		return types.Binary{}, types.Binary{}, err
-	}
+		// {field: {$bitsAllClear: BinData()}}
+		for b := 0; b < len(mask.B); b++ {
+			byteAt := mask.B[b]
+
+			if byteAt == 0 {
+				continue
+			}
+
+			if b < 8 {
+				bitmask |= uint64(byteAt) << b
+			} else {
+				bitmask |= 1 << 63
+			}
+		}
```

Comment:
Let's remove that code if it is not tested (and, I guess, does not work the same way as MongoDB)

Reply:
I've added tests for that code.